src/PolyType.SourceGenerator/Parser/Parser.ModelMapper.cs

```diff
@@ -16,13 +18,93 @@ private TypeShapeModel MapModel(TypeDataModel model, TypeId typeId, string sourc
+        ITypeSymbol[] typeArgs = namedType?.GetRecursiveTypeArguments() ?? [];
+        foreach ((INamedTypeSymbol openType, IAssemblySymbol associatedAssembly, Location? location) in associatedTypeSymbols)
+        {
+            if (!SymbolEqualityComparer.Default.Equals(openType.ContainingAssembly, associatedAssembly))
+            {
+                ReportDiagnostic(AssociatedTypeInExternalAssembly, location, openType.GetFullyQualifiedName());
+                continue;
+            }
+
+            if (!this.IsAccessibleSymbol(openType))
+            {
+                // Skip types that are not accessible in the current scope
+                ReportDiagnostic(AssociatedTypeInaccessibleError, location, openType.GetFullyQualifiedName());
+                continue;
+            }
+
+            IMethodSymbol? defaultCtor = null;
+            if (!openType.IsUnboundGenericType)
+            {
+                if (TryGetCtorOrReport(openType, out defaultCtor))
+                {
+                    associatedTypesBuilder.Add(CreateAssociatedTypeId(openType, openType));
+                }
+                else
+                {
+                    continue;
+                }
+            }
+            else
+            {
+                if (openType.OriginalDefinition.ConstructRecursive(typeArgs) is INamedTypeSymbol closedType)
+                {
+                    if (TryGetCtorOrReport(closedType, out defaultCtor))
+                    {
+                        associatedTypesBuilder.Add((CreateAssociatedTypeId(openType, closedType)));
+                    }
+                    else
+                    {
+                        continue;
+                    }
+                }
+                else if (openType.Arity != typeArgs.Length)
+                {
+                    ReportDiagnostic(AssociatedTypeArityMismatch, location, openType.GetFullyQualifiedName(), openType.Arity, typeArgs.Length);
+                    continue;
+                }
+            }
+
+            if (openType.DeclaredAccessibility != Accessibility.Public || defaultCtor?.DeclaredAccessibility != Accessibility.Public)
+            {
+                ReportDiagnostic(AssociatedTypeInternal, location, openType.GetFullyQualifiedName());
+            }
+
+            bool TryGetCtorOrReport(INamedTypeSymbol type, out IMethodSymbol? defaultCtor)
+            {
+                defaultCtor = type.InstanceConstructors.FirstOrDefault(c => c.Parameters.IsEmpty);
+                if (defaultCtor is null || !this.IsAccessibleSymbol(defaultCtor))
+                {
+                    ReportDiagnostic(AssociatedTypeInaccessibleError, location, openType.GetFullyQualifiedName());
+                    return false;
+                }
+
+                return true;
+            }
+        }
+
+        ImmutableEquatableArray<AssociatedTypeId> associatedTypes = associatedTypesBuilder.ToImmutableEquatableArray();
```

Comment:
Feels like this could all be extracted into a helper function returning the associated types.